File: Medium/966. Vowel Spellchecker/solution.cs
public class Solution {
    public string[] Spellchecker(string[] wordlist, string[] queries) {
        HashSet<string> words = new HashSet<string>(wordlist);

        Dictionary<string, string> lowerCase = new Dictionary<string, string>();
        Dictionary<string, string> vowelError = new Dictionary<string, string>();

        foreach(string word in wordlist){
            string lower = word.ToLower();
            if(!lowerCase.ContainsKey(lower)){
                lowerCase[lower] = word;
            }

            string pattern = pattern_word(word.ToLower());
            if(!vowelError.ContainsKey(pattern)){
                vowelError[pattern] = word;
            }
        }

        List<string> res = new List<string>();
        foreach(string query in queries){
            if(words.Contains(query)){
                res.Add(query);
            } else if(lowerCase.ContainsKey(query.ToLower())){
                res.Add(lowerCase[query.ToLower()]);
            } else {
                string pattern = pattern_word(query.ToLower());
                if(vowelError.ContainsKey(pattern)){
                    res.Add(vowelError[pattern]);
                } else {
                    res.Add("");
                }
            }
        }

        return res.ToArray();
    }

    public string pattern_word(string word){
        char[] arr = word.ToCharArray();
        for(int i = 0; i < arr.Length; i++){
            if("aeiou".Contains(arr[i])){
                arr[i] = '_';
            }
        }
        return new string(arr);
    }
}

File: Medium/966. Vowel Spellchecker/solution.py
class Solution:
    def spellchecker(self, wordlist: List[str], queries: List[str]) -> List[str]:
        words = set(wordlist)
        lower = {}
        pattern = {}

        for i in wordlist:
            wordlower = i.lower()
            if wordlower not in lower:
                lower[wordlower] = i

            patternword = self.pattern(i.lower())
            if patternword not in pattern:
                pattern[patternword] = i

        res = []
        for i in queries:
            if i in words:
                res.append(i)
            elif i.lower() in lower:
                res.append(lower[i.lower()])
            elif self.pattern(i.lower()) in pattern:
                res.append(pattern[self.pattern(i.lower())])
            else:
                res.append("")
        return res

    def pattern(self, word):
        return "".join("_" if i in "aiuoe" else i for i in word)
